File: cpp/tool/Verifier5.cpp
/**
 * build with command:
 * g++ Verifier5.cpp  -lcrypto -std=c++0x
 */
#include "../src/DynamicKey5.h"
#include <iostream>
#include <cstdint>
using namespace agora::tools;

template<typename T>
inline std::string map_to_string(const T & c, const std::string& separator)
{
    std::ostringstream oss;
    std::string s = "";
    for (const auto & v: c) {
        oss << s << v.first << ":" << v.second;
        s = separator;
    }

    return oss.str();
}

int main(int argc, char const *argv[]) {
    if (argc < 6) {
        std::cout << argv[0] << " appID appCertificate channelName uid key" << std::endl;
        return 0;
    }

    std::string appID = argv[1];
    std::string appCertificate = argv[2];
    std::string channelName = argv[3];
    uint32_t uid = strtoul(argv[4], NULL, 10);
    std::string keyString = argv[5];

    DynamicKey5 key5;
    key5.fromString(keyString);

    std::cout << std::endl;
    std::cout << "Service     " << key5.serviceType << std::endl;
    std::cout << "App ID:     " << key5.appID << std::endl;
    std::cout << "unixTs:     " << key5.unixTs << std::endl;
    std::cout << "expiredTs:  " << key5.expiredTs << std::endl;
    std::cout << "Extra:      [" << map_to_string(key5.extra, ", ") << "]" << std::endl;
    std::cout << "randomInt:  " << key5.randomInt << std::endl;
    std::cout << "Signature:  " << key5.signature << std::endl;
    std::cout << std::endl;

    std::cout << std::endl;
    std::cout << "Original \t " << keyString << std::endl;
    std::cout << "Uid = 0 \t " << DynamicKey5::generateMediaChannelKey(appID, appCertificate, channelName, key5.unixTs, key5.randomInt, 0, key5.expiredTs) << std::endl;
    std::cout << "Uid = " << uid << " \t " << DynamicKey5::generateMediaChannelKey(appID, appCertificate, channelName, key5.unixTs, key5.randomInt, uid, key5.expiredTs) << std::endl;

    return 0;
}

File: java/tool/io/agora/media/tool/Verifier5.java
package io.agora.media.tool;

import io.agora.media.DynamicKey5;
import org.apache.commons.codec.binary.Hex;

/**
 * Created by liwei on 8/2/17.
 */
public class Verifier5 {
    public static void main(String[] args) throws Exception {
        if (args.length < 5) {
            System.out.println("java io.agora.media.tool.Verifier5 appID appCertificate channelName uid channelKey");
            return;
        }

        String appID = args[0];
        String appCertificate = args[1];
        String channelName = args[2];
        int uid = Integer.parseInt(args[3]);
        String channelKey = args[4];

        DynamicKey5 key5 = new DynamicKey5();
        if (! key5.fromString(channelKey)) {
            System.out.println("Faile to parse key");
            return ;
        }

        System.out.println("signature " + key5.content.signature);
        System.out.println("appID " + new String(Hex.encodeHex(key5.content.appID, false)));
        System.out.println("unixTs " + key5.content.unixTs);
        System.out.println("randomInt " + key5.content.salt);
        System.out.println("expiredTs " + key5.content.expiredTs);
        System.out.println("extra " + key5.content.extra.size());
        System.out.println("service " + key5.content.serviceType);

        System.out.println();
        System.out.println("Original \t\t " + channelKey);
        System.out.println("Uid = 0 \t\t " + DynamicKey5.generateMediaChannelKey(appID, appCertificate, channelName, key5.content.unixTs, key5.content.salt, 0, key5.content.expiredTs));
        System.out.println("Uid =  " + uid + " \t " + DynamicKey5.generateMediaChannelKey(appID, appCertificate, channelName, key5.content.unixTs, key5.content.salt, uid, key5.content.expiredTs));

        String signature = DynamicKey5.generateSignature(appCertificate,
                key5.content.serviceType,
                appID,
                key5.content.unixTs,
                key5.content.salt,
                channelName,
                uid,
                0,
                key5.content.extra
        );
        System.out.println("generated signature " + signature);
    }
}
